docs/design/remerge/rfc.md

```diff
@@ -0,0 +1,938 @@
+# Remerge: A syncable store for generic data types
+
+Unfortunately, with the (indefinite) "pause" of Mentat, there's no obvious path
+forward for new synced data types beyond 'the Sync team implements a new
+component'. Presumably, at some point we decided this was both desirable, but
+unworkable, hence designing Mentat.
+
+After some thought, I've come up with a design that gets us some of the benefits
+of Mentat with the following major benefits (compared to Mentat)
+
+- Works on top of Sync 1.5, including interfacing with existing collections,
+- Doesn't change the sync data model substantially.
+- Has storage which is straightforward to implement on top of SQLite.
+
+Because the system needs a name, I've decided to call it Remerge.
+
+## Background/Goals/Etc
+
+In one of the AllHands, Lina had a presentation which defined three different
+types of sync data stores.
+
+1. Tree stores (bookmarks). The defining features of these stores are that:
+    - They represent a tree.
+    - They are considered corrupt if tree constraints are invalidated.
+2. Log stores (history). The defining features of these stores are that:
+    - Typically too large to fit in memory.
+    - We expect to only sync a subset of the records in them.
+3. Record stores (logins, addresses, credit cards, addons, etc)
+
+This document describes a plan for syncing "Type 3" data stores in a generic
+way, however extended to allow the following additional major features not
+present in the current system:
+
+1. Schema evolution with both forwards and backwards compatibility, with the
+   ability to break compatibility gradually, avoiding breaking sync with
+   versions that are still supported
+2. Flexible merge logic.
+3. Relatively rich data types, with the ability to add more data types over time
+   without breaking old clients (which IIUC was a problem mentat had).
+
+### Main Ideas
+
+The high level summary of the "big ideas" behind Remerge, in very rough order of
+importance, are as follows:
+
+1. By syncing the schema, we can allow older clients to understand enough
+   semantic data about the record format so that they're still able to sync and
+   perform correct three way merges, even if they don't actually understand how
+   to use the data that they're merging.
```

Comment:
This would be a pretty big undertaking, and I don't think it's worth it; just throwing the idea out there...what if we stored all available storage schemas on the server, at a public URI (`info/schemas`?), and stored the client IDs, vector clocks, and minimum and current schema versions in `meta/global`?

That would give us more centralized control over schema handling (to prevent clients from doing potentially silly things), and let us eventually deprecate old schema versions on the server. It would also reduce the surface area for collections that we need to update, since we already fetch and update `m/g` as part of the sync.

(I never imagined I'd suggest XML namespaces, but here we are 😂)